File: C++/topological_sort.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

/*
╔════════════════════════════════════════════════════════╗
║         Topological Sort (Kahn's Algorithm)           ║
╠════════════════════════════════════════════════════════╣
║ Description:                                           ║
║ 1) Compute indegree of each node.                     ║
║ 2) Enqueue all zero‑indegree nodes.                   ║
║ 3) Pop, add to order, decrement neighbors' indegree. ║
║ 4) Enqueue new zero‑indegree until done.              ║
╠════════════════════════════════════════════════════════╣
║ Flow Diagram (ASCII):                                  ║
║ Graph:                                                 ║
║ 0→1,0→2; 1→3;2→3,2→4;3→5;4→5                            ║
║ indegree: [0,1,1,2,1,2]                                ║
║ Queue init: [0] → order=[ ]                            ║
║ Pop 0 → order=[0] → indegree[1]-- →0 , [2]-- →0 → enqueue[1,2] ║
║ Pop 1 → order=[0,1] → indegree[3]-- →1 → …             ║
║ → Final order: [0,1,2,3,4,5]                           ║
╚════════════════════════════════════════════════════════╝
*/

vector<int> topoSort(int V, const vector<vector<int>>& adj) {
    vector<int> indegree(V, 0), order;
    queue<int> q;

    // 1) Count indegrees
    for (int u = 0; u < V; ++u)
        for (int v : adj[u])
            ++indegree[v];

    // 2) Enqueue zeros
    for (int i = 0; i < V; ++i)
        if (indegree[i] == 0)
            q.push(i);

    // 3) Process
    while (!q.empty()) {
        int u = q.front(); q.pop();
        order.push_back(u);
        for (int v : adj[u]) {
            if (--indegree[v] == 0)
                q.push(v);
        }
    }
    return order;
}

int main() {
    int V = 6;
    vector<vector<int>> graph = {
        {1,2}, {3}, {3,4}, {5}, {5}, {}
    };
    auto order = topoSort(V, graph);
    for (int u : order)
        cout << u << " ";
    cout << "\n";
    return 0;
} 

File: Go/topological_sort.go
/*
╔════════════════════════════════════════════════════════╗
║           Topological Sort (Kahn's Algorithm)         ║
╠════════════════════════════════════════════════════════╣
║ Description:                                           ║
║ Sort vertices in a DAG such that for every directed   ║
║ edge u→v, u comes before v. Use in-degree count and  ║
║ queue to process nodes with no incoming edges.        ║
╠════════════════════════════════════════════════════════╣
║ Flow Diagram (ASCII):                                  ║
║   Graph: 5→2→3→1                                      ║
║           ↓   ↑                                       ║
║           4 → 0                                       ║
║                                                      ║
║   in-degree: [1,1,1,0,0]                             ║
║   queue: [4,5] → process 4 → queue: [5,0]            ║
║   queue: [5,0] → process 5 → queue: [0,2]            ║
║   queue: [0,2] → process 0 → queue: [2,3]            ║
║   queue: [2,3] → process 2 → queue: [3]              ║
║   queue: [3] → process 3 → queue: [1]                ║
║   queue: [1] → process 1 → queue: []                 ║
║                                                      ║
║   Result: [4,5,0,2,3,1]                             ║
╚════════════════════════════════════════════════════════╝
*/

package main

import (
    "container/list"
    "fmt"
)

func topologicalSort(numVertices int, edges [][]int) []int {
    // Build adjacency list and in-degree count
    adj := make([][]int, numVertices)
    inDegree := make([]int, numVertices)

    for _, edge := range edges {
        u, v := edge[0], edge[1]
        adj[u] = append(adj[u], v)
        inDegree[v]++
    }

    // Initialize queue with vertices of in-degree 0
    queue := list.New()
    for i := 0; i < numVertices; i++ {
        if inDegree[i] == 0 {
            queue.PushBack(i)
        }
    }

    result := make([]int, 0, numVertices)
    for queue.Len() > 0 {
        u := queue.Remove(queue.Front()).(int)
        result = append(result, u)

        // Decrease in-degree of neighbors
        for _, v := range adj[u] {
            inDegree[v]--
            if inDegree[v] == 0 {
                queue.PushBack(v)
            }
        }
    }

    // Check for cycle
    if len(result) != numVertices {
        return nil // Cycle detected
    }

    return result
}

func main() {
    numVertices := 6
    edges := [][]int{
        {5, 2},
        {5, 0},
        {4, 0},
        {4, 1},
        {2, 3},
        {3, 1},
    }

    result := topologicalSort(numVertices, edges)
    if result == nil {
        fmt.Println("Cycle detected in graph")
    } else {
        fmt.Println("Topological order:", result)
    }
} 